File: gallery/pdf_writer/src/jpeg/bin/gallery/pdf_writer/src/jpeg/HuffmanDecoder_new.js
class HuffmanTable  {
  constructor() {
    this.bits = new Int32Array(16);
    this.values = [];
    this.maxCode = new Int32Array(16);
    this.minCode = new Int32Array(16);
    this.valPtr = new Int32Array(16);
    this.tableClass = 0;
    this.tableId = 0;
    let i = 0;
    while (i < 16) {
      this.bits[i] = 0;
      this.maxCode[i] = -1;
      this.minCode[i] = 0;
      this.valPtr[i] = 0;
      i = i + 1;
    };
  }
  build () {
    let code = 0;
    let valueIdx = 0;
    let i = 0;
    while (i < 16) {
      const count = this.bits[i];
      if ( count > 0 ) {
        this.minCode[i] = code;
        this.valPtr[i] = valueIdx;
        valueIdx = valueIdx + count;
        code = code + count;
        this.maxCode[i] = code - 1;
      } else {
        this.maxCode[i] = -1;
        this.minCode[i] = 0;
        this.valPtr[i] = valueIdx;
      }
      code = (code << 1);
      i = i + 1;
    };
  };
  decode (reader) {
    let code = 0;
    let length = 0;
    while (length < 16) {
      const bit = reader;
      code = (((code << 1)) | bit);
      const maxC = this.maxCode[length];
      if ( maxC >= 0 ) {
        if ( code <= maxC ) {
          const minC = this.minCode[length];
          const ptr = this.valPtr[length];
          const idx = ptr + (code - minC);
          return this.values[idx];
        }
      }
      length = length + 1;
    };
    console.log("Huffman decode error: code not found");
    return 0;
  };
  resetArrays () {
    let i = 0;
    while (i < 16) {
      this.bits[i] = 0;
      this.maxCode[i] = -1;
      this.minCode[i] = 0;
      this.valPtr[i] = 0;
      i = i + 1;
    };
    this.values.length = 0;
  };
}
class HuffmanDecoder  {
  constructor() {
    this.dcTable0 = new HuffmanTable();
    this.dcTable1 = new HuffmanTable();
    this.acTable0 = new HuffmanTable();
    this.acTable1 = new HuffmanTable();
  }
  getDCTable (id) {
    if ( id == 0 ) {
      return this.dcTable0;
    }
    return this.dcTable1;
  };
  getACTable (id) {
    if ( id == 0 ) {
      return this.acTable0;
    }
    return this.acTable1;
  };
  parseDHT (data, pos, length) {
    const endPos = pos + length;
    while (pos < endPos) {
      const tableInfo = data._view.getUint8(pos);
      pos = pos + 1;
      const tableClass = (tableInfo >> 4);
      const tableId = (tableInfo & 15);
      let table = this.getDCTable(tableId);
      if ( tableClass == 1 ) {
        table = this.getACTable(tableId);
      }
      table.tableClass = tableClass;
      table.tableId = tableId;
      table.resetArrays();
      let totalSymbols = 0;
      let i = 0;
      while (i < 16) {
        const count = data._view.getUint8(pos);
        table.bits[i] = count;
        totalSymbols = totalSymbols + count;
        pos = pos + 1;
        i = i + 1;
      };
      i = 0;
      while (i < totalSymbols) {
        table.values.push(data._view.getUint8(pos));
        pos = pos + 1;
        i = i + 1;
      };
      table.build();
      let classStr = "DC";
      if ( tableClass == 1 ) {
        classStr = "AC";
      }
      console.log((((("  Huffman table " + classStr) + ((tableId.toString()))) + ": ") + ((totalSymbols.toString()))) + " symbols");
    };
  };
}

File: gallery/pdf_writer/src/jpeg/bin/gallery/pdf_writer/src/jpeg/HuffmanDecoder_new.go
package main
import (
  "fmt"
  "strconv"
)

type GoNullable struct {
  value interface{}
  has_value bool
}

type HuffmanTable struct { 
  bits []int64 `json:"bits"` 
  values []int64 `json:"values"` 
  maxCode []int64 `json:"maxCode"` 
  minCode []int64 `json:"minCode"` 
  valPtr []int64 `json:"valPtr"` 
  tableClass int64 `json:"tableClass"` 
  tableId int64 `json:"tableId"` 
}

func CreateNew_HuffmanTable() *HuffmanTable {
  me := new(HuffmanTable)
  me.bits = 
  make([]int64, int64(16))
  
  me.values = make([]int64,0)
  me.maxCode = 
  make([]int64, int64(16))
  
  me.minCode = 
  make([]int64, int64(16))
  
  me.valPtr = 
  make([]int64, int64(16))
  
  me.tableClass = int64(0)
  me.tableId = int64(0)
  var i int64= int64(0);
  for i < int64(16) {
    me.bits[i] = int64(0)
    me.maxCode[i] = int64(-1)
    me.minCode[i] = int64(0)
    me.valPtr[i] = int64(0)
    i = i + int64(1); 
  }
  return me;
}
func (this *HuffmanTable) build () () {
  var code int64= int64(0);
  var valueIdx int64= int64(0);
  var i int64= int64(0);
  for i < int64(16) {
    var count int64= this.bits[i];
    if  count > int64(0) {
      this.minCode[i] = code
      this.valPtr[i] = valueIdx
      valueIdx = valueIdx + count; 
      code = code + count; 
      this.maxCode[i] = code - int64(1)
    } else {
      this.maxCode[i] = int64(-1)
      this.minCode[i] = int64(0)
      this.valPtr[i] = valueIdx
    }
    code = int64(code << uint(int64(1))); 
    i = i + int64(1); 
  }
}
func (this *HuffmanTable) decode (reader *BitReader) int64 {
  var code int64= int64(0);
  var length int64= int64(0);
  for length < int64(16) {
    var bit int64= reader;
    code = int64((int64(code << uint(int64(1)))) | bit); 
    var maxC int64= this.maxCode[length];
    if  maxC >= int64(0) {
      if  code <= maxC {
        var minC int64= this.minCode[length];
        var ptr int64= this.valPtr[length];
        var idx int64= ptr + (code - minC);
        return this.values[idx]
      }
    }
    length = length + int64(1); 
  }
  fmt.Println( "Huffman decode error: code not found" )
  return int64(0)
}
func (this *HuffmanTable) resetArrays () () {
  var i int64= int64(0);
  for i < int64(16) {
    this.bits[i] = int64(0)
    this.maxCode[i] = int64(-1)
    this.minCode[i] = int64(0)
    this.valPtr[i] = int64(0)
    i = i + int64(1); 
  }
  this.values = this.values[:0]
}
type HuffmanDecoder struct { 
  dcTable0 *HuffmanTable `json:"dcTable0"` 
  dcTable1 *HuffmanTable `json:"dcTable1"` 
  acTable0 *HuffmanTable `json:"acTable0"` 
  acTable1 *HuffmanTable `json:"acTable1"` 
}

func CreateNew_HuffmanDecoder() *HuffmanDecoder {
  me := new(HuffmanDecoder)
  me.dcTable0 = CreateNew_HuffmanTable()
  me.dcTable1 = CreateNew_HuffmanTable()
  me.acTable0 = CreateNew_HuffmanTable()
  me.acTable1 = CreateNew_HuffmanTable()
  return me;
}
func (this *HuffmanDecoder) getDCTable (id int64) *HuffmanTable {
  if  id == int64(0) {
    return this.dcTable0
  }
  return this.dcTable1
}
func (this *HuffmanDecoder) getACTable (id int64) *HuffmanTable {
  if  id == int64(0) {
    return this.acTable0
  }
  return this.acTable1
}
func (this *HuffmanDecoder) parseDHT (data []byte, pos int64, length int64) () {
  var endPos int64= pos + length;
  for pos < endPos {
    var tableInfo int64= int64(data[pos]);
    pos = pos + int64(1); 
    var tableClass int64= int64(tableInfo >> uint(int64(4)));
    var tableId int64= int64(tableInfo & int64(15));
    var table *HuffmanTable= this.getDCTable(tableId);
    if  tableClass == int64(1) {
      table = this.getACTable(tableId); 
    }
    table.tableClass = tableClass; 
    table.tableId = tableId; 
    table.resetArrays();
    var totalSymbols int64= int64(0);
    var i int64= int64(0);
    for i < int64(16) {
      var count int64= int64(data[pos]);
      table.bits[i] = count
      totalSymbols = totalSymbols + count; 
      pos = pos + int64(1); 
      i = i + int64(1); 
    }
    i = int64(0); 
    for i < totalSymbols {
      table.values = append(table.values,int64(data[pos])); 
      pos = pos + int64(1); 
      i = i + int64(1); 
    }
    table.build();
    var classStr string= "DC";
    if  tableClass == int64(1) {
      classStr = "AC"; 
    }
    fmt.Println( (((("  Huffman table " + classStr) + (strconv.FormatInt(tableId, 10))) + ": ") + (strconv.FormatInt(totalSymbols, 10))) + " symbols" )
  }
}
